go/vt/vttablet/tabletserver/repltracker/writer.go

```diff
@@ -238,14 +240,63 @@ func (w *heartbeatWriter) enableWrites(enable bool) {
 			w.ticks.Start(w.writeHeartbeat)
 		}()
 	case false:
-		w.ticks.Stop()
+		// We stop the ticks in a separate go routine because it can block if the write is stuck on semi-sync ACKs.
+		// At the same time we try and kill the write that is in progress. We use the context and its cancellation
+		// for coordination between the two go-routines. In the end we will have guaranteed that the ticks have stopped
+		// and no write is in progress.
+		ctx, cancel := context.WithCancel(context.Background())
+		go func() {
+			w.ticks.Stop()
+			cancel()
+		}()
+		w.killWritesUntilStopped(ctx)
+
 		if w.onDemandDuration > 0 {
 			// Let the next RequestHeartbeats() go through
 			w.allowNextHeartbeatRequest()
 		}
 	}
 }
 
+// killWritesUntilStopped tries to kill the write in progress until the ticks have stopped.
+func (w *heartbeatWriter) killWritesUntilStopped(ctx context.Context) {
+	for {
+		select {
+		case <-ctx.Done():
+			// If the context has been cancelled, then we know that the ticks have stopped.
+			// This guarantees that there are no writes in progress, so there is nothing to kill.
+			return
+		default:
+		}
+		// Actually try to kill the query.
+		err := w.killWrite()
+		w.recordError(err)
+		// We sleep for 100 millisecond, to prevent killing the query over and over very quickly.
+		time.Sleep(100 * time.Millisecond)
+	}
```

Comment:
Let's use the ticker:
```suggestion
	for {
		// Actually try to kill the query.
		err := w.killWrite()
		w.recordError(err)
		select {
		case <-ctx.Done():
			// If the context has been cancelled, then we know that the ticks have stopped.
			// This guarantees that there are no writes in progress, so there is nothing to kill.
			return
		case <-ticker.C:
		}
	}
```